### PR TITLE
adding user-specific budgets

### PR DESCRIPTION
**Pull request by krrishdholakia**
Following from our conversation on Discord. 

This is v1 of a budget manager to track spend per model. 

It currently stores this in-memory. We would need to move this to storing it in a local file or ideally a hosted db (can expose + manage that via `api.litellm.ai` if you'd prefer).

### CODE REVIEW

**Comment by krrishdholakia** [comment]
It seems like i'm failing some ci work? Is there any steps i can take to test and run this locally? @yujonglee

**Comment by yujonglee** [comment]
@krrishdholakia  run `poetry lock \[--no-update\]` or `task sync` please!

**Comment by krrishdholakia** [comment]
Ok - sounds good, thanks

**Comment by krrishdholakia** [comment]
Hmm, Running `poetry lock \[--no-update\]` threw errors for me.

So i ran `poetry lock --no-update` and it still failed.

File: proxy/main.py [hunk 1]
```diff
@@ -19,38 +19,60 @@ async def health(response: Response):
...
+    total_budget = data["total_budget"]
     api_key = f"sk-fastrepl-{secrets.token_urlsafe(16)}"
 
+    print(f"config: {config}")
+
+    
     # Append the new API key to the existing list in the Auth_Token variable
-    tokens_string = os.getenv("AUTH_TOKEN", "")
-    auth_tokens = [tokens_string] if tokens_string != "" else []
+    tokens_string = config.get("AUTH_TOKEN", "")
```
Comment (yujonglee) [inline proxy/main.py:56]: Let's annotate this `: str` to make `mypy` happy

File: proxy/main.py [hunk 2]
```diff
@@ -19,38 +19,60 @@ async def health(response: Response):
 
 @app.post("/chat/completions", status_code=status.HTTP_200_OK)
 async def completion(request: Request, response: Response):
-    tokens_string = os.getenv("AUTH_TOKEN", "")
-    tokens = tokens_string.split(",")
+    tokens_string = config.get("AUTH_TOKEN", "")
+    tokens = tokens_string.split(',') if tokens_string != "" else []
```
Comment (yujonglee) [inline proxy/main.py:23]: Let's annotate `: str` to make `mypy` happy

**Review by yujonglee — CHANGES_REQUESTED** [review]
@krrishdholakia Sorry for the inconvenience.

Your poetry lock update did pass the install dependency stage, but now it failed on lint stage.

See `Taskfile.yaml`. `lint` and `fmt` should pass.

If you cannot make it to pass lint, you can use `# type: ignore` and move forward. I can handle that later.

**Comment by krrishdholakia** [comment]
No your fine - i didn't catch that. Thanks for that. I'll run linting before pushing.

File: proxy/main.py [hunk 3]
```diff
@@ -19,38 +19,60 @@ async def health(response: Response):
...
+    return llm.completion(**data, api_key=api_key, budget_manager=budget_manager)
 
 
-@app.get("/key/new", status_code=status.HTTP_200_OK)
-async def generate_key(response: Response):
+@app.post("/key/new", status_code=status.HTTP_200_OK)
+async def generate_key(request: Request, response: Response):
+    try:
+        data = await request.json() 
+        total_budget = data["total_budget"]
```
Comment (yujonglee) [inline proxy/main.py:44]: I think we can remove this

File: proxy/main.py [hunk 4]
```diff
@@ -19,38 +19,60 @@ async def health(response: Response):
...
+    except:
+        response.status_code = status.HTTP_400_BAD_REQUEST
+        return
+    
+    total_budget = data["total_budget"]
     api_key = f"sk-fastrepl-{secrets.token_urlsafe(16)}"
 
+    print(f"config: {config}")
```
Comment (yujonglee) [inline proxy/main.py:52]: We can remove this

File: proxy/main.py [hunk 5]
```diff
@@ -19,38 +19,60 @@ async def health(response: Response):
...
-    load_dotenv()  # reload the values into the .env
-    return {"api_key": api_key}
+
+    # Update the dynamic variable containing the auth_token list
+    config["AUTH_TOKEN"] = ",".join(auth_tokens)
+
+    # initialize a user budget to track spend 
+    try:
+        response = budget_manager.create_budget(total_budget=total_budget, user=api_key)
+        print(f"budget response: {response}")
```
Comment (yujonglee) [inline proxy/main.py:69]: we can remove this :)

File: proxy/llm.py [hunk 6]
```diff
@@ -100,12 +100,18 @@ def completion(**kwargs) -> ModelResponse:
...
-            if overide_model is not None:
-                kwargs["model"] = overide_model
-            return litellm.completion(**kwargs)
+            if override_model is not None:
+                kwargs["model"] = override_model
+            if budget_manager.get_current_cost(user=api_key) < budget_manager.get_total_budget(user=api_key):
+                response = litellm.completion(**kwargs)
+                budget_manager.update_cost(completion_obj=response, user=api_key)
+            else:
+                raise Exception("No Budget - sorry") # maybe we can raise a 429 rate-limit exception here instead
```
Comment (yujonglee) [inline proxy/llm.py:113]: Yes -  let's raise 429

**Review by yujonglee — CHANGES_REQUESTED** [review]

File: proxy/main.py (hunk 3 above)
Comment (krrishdholakia) [reply proxy/main.py:44]: how would you set user-specific budgets? Would it be the same for all? 

I assumed you'd have different tiers (free, paid, etc.)

**Comment by yujonglee** [comment]
Working on this in #5(https://github.com/fastrepl/proxy/pull/6/commits/9bbfe528fa880f329a1604cf3c52f147e0f0cfbe).

Your initial PR helped me as a reference.

**Pull request closed without merging**